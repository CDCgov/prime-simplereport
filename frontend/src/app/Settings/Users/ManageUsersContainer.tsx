import React from "react";
import { gql, useQuery, useMutation } from "@apollo/client";
import { useSelector } from "react-redux";

import { UserRole, UserPermission, Role } from "../../permissions";

import ManageUsers from "./ManageUsers";

const GET_USERS = gql`
  query GetUsers {
    users {
      id
      firstName
      middleName
      lastName
      email
    }
  }
`;

export const GET_USER = gql`
  query GetUser($id: ID!) {
    user(id: $id) {
      id
      firstName
      middleName
      lastName
      roleDescription
      role
      permissions
      email
      status
      organization {
        testingFacility {
          id
          name
        }
      }
    }
  }
`;

// structure for `getUser` query
export interface SettingsUser {
  id: string;
  firstName: string;
  middleName: string;
  lastName: string;
  roleDescription: UserRole;
  role: Role;
  permissions: UserPermission[];
  email: string;
  status: string;
  organization: {
    testingFacility: UserFacilitySetting[];
  };
}

// structure for `getUsers` query
export interface LimitedUser {
  id: string;
  firstName: string;
  middleName: string;
  lastName: string;
  email: string;
}

interface UserData {
  users: LimitedUser[];
}

export interface SingleUserData {
  user: SettingsUser;
}

const UPDATE_USER_PRIVILEGES = gql`
  mutation UpdateUserPrivileges(
    $id: ID!
    $role: Role!
    $accessAllFacilities: Boolean!
    $facilities: [ID!]!
  ) {
    updateUserPrivileges(
      id: $id
      role: $role
      accessAllFacilities: $accessAllFacilities
      facilities: $facilities
    ) {
      id
    }
  }
`;

const DELETE_USER = gql`
  mutation SetUserIsDeleted($id: ID!, $deleted: Boolean!) {
    setUserIsDeleted(id: $id, deleted: $deleted) {
      id
    }
  }
`;

const REACTIVATE_USER = gql`
  mutation ReactivateUser($id: ID!) {
    reactivateUser(id: $id) {
      id
    }
  }
`;

const ADD_USER_TO_ORG = gql`
  mutation AddUserToCurrentOrg(
    $firstName: String
    $lastName: String!
    $email: String!
    $role: Role!
  ) {
    addUserToCurrentOrg(
      firstName: $firstName
      lastName: $lastName
      email: $email
      role: $role
    ) {
      id
    }
  }
`;

const GET_FACILITIES = gql`
  query GetFacilitiesForManageUsers {
    organization {
      testingFacility {
        id
        name
      }
    }
  }
`;

interface FacilityData {
  organization: {
    testingFacility: UserFacilitySetting[];
  };
}

export interface UserFacilitySetting {
  id: string;
  name: string;
}

export interface NewUserInvite {
  firstName: string;
  lastName: string;
  email: string;
  role: UserRole | string | undefined; // TODO: clean this up or delete it if we are not supporting this feature
}

const ManageUsersContainer: any = () => {
  const loggedInUser = useSelector((state) => (state as any).user as User);
  const [updateUserPrivileges] = useMutation(UPDATE_USER_PRIVILEGES);
  const [deleteUser] = useMutation(DELETE_USER);
  const [reactivateUser] = useMutation(REACTIVATE_USER);
  const [addUserToOrg] = useMutation(ADD_USER_TO_ORG);

  const { data, loading, error, refetch: getUsers } = useQuery<UserData, {}>(
    GET_USERS,
    { fetchPolicy: "no-cache" }
  );

  const {
    data: dataFacilities,
    loading: loadingFacilities,
    error: errorFacilities,
  } = useQuery<FacilityData, {}>(GET_FACILITIES, {
    fetchPolicy: "no-cache",
  });

  if (loading || loadingFacilities) {
    return <p> Loading... </p>;
  }

  if (error || errorFacilities) {
    throw error || errorFacilities;
  }

  if (data === undefined) {
    return <p>Error: Users not found</p>;
  }

  if (dataFacilities === undefined) {
    return <p>Error: Facilities not found</p>;
  }

  const allFacilities = dataFacilities.organization
    .testingFacility as UserFacilitySetting[];

  return (
    <ManageUsers
      users={data.users}
      loggedInUser={loggedInUser}
      allFacilities={allFacilities}
      updateUserPrivileges={updateUserPrivileges}
      addUserToOrg={addUserToOrg}
      deleteUser={deleteUser}
      reactivateUser={reactivateUser}
      getUsers={getUsers}
    />
  );
};

export default ManageUsersContainer;
