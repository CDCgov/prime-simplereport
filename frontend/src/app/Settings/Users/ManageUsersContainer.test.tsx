import React from "react";
import { MemoryRouter } from "react-router-dom";
import { Provider } from "react-redux";
import { MockedProvider, MockedResponse } from "@apollo/client/testing";
import {
  act,
  render,
  screen,
  waitFor,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import configureStore from "redux-mock-store";
import { AnyAction, Store } from "redux";
import userEvent from "@testing-library/user-event";
import { cloneDeep } from "lodash";

import {
  GetUserDocument,
  GetUsersAndStatusDocument,
} from "../../../generated/graphql";
import {
  ORG_ADMIN_REACTIVATE_COPY,
  SITE_ADMIN_REACTIVATE_COPY,
} from "../../commonComponents/UserDetails/ReactivateUserModal";

import ManageUsersContainer from "./ManageUsersContainer";

describe("ManageUsersContainer", () => {
  const mockStore = configureStore([]);
  const storeData = {
    organization: {
      name: "Organization Name",
    },
    user: {
      firstName: "Kim",
      lastName: "Mendoza",
      roleDescription: "Admin user",
      isAdmin: false,
    },
    facilities: [
      { id: "1", name: "Facility 1" },
      { id: "2", name: "Facility 2" },
    ],
  };
  const mockedUsersWithStatus = [
    {
      id: "3a4a221d-a8ca-42b3-aa03-37b93266025b",
      firstName: "Ben",
      middleName: "Billy",
      lastName: "Barnes",
      email: "ben@example.com",
      status: "ACTIVE",
      __typename: "ApiUserWithStatus",
    },
    {
      id: "1029653e-24d9-428e-83b0-468319948902",
      firstName: "Bob",
      middleName: null,
      lastName: "Bobberoo",
      email: "bob@example.com",
      status: "ACTIVE",
      __typename: "ApiUserWithStatus",
    },
    {
      id: "60bb9e3a-fe8a-4b81-b894-b01649c95e70",
      firstName: "Jamar",
      middleName: "Donald",
      lastName: "Jackson",
      email: "jamar@example.com",
      status: "ACTIVE",
      __typename: "ApiUserWithStatus",
    },
    {
      id: "0d3fa224-3d56-4382-b89c-9d8e415e59b3",
      firstName: "Ruby",
      middleName: "Raven",
      lastName: "Reynolds",
      email: "ruby@example.com",
      status: "ACTIVE",
      __typename: "ApiUserWithStatus",
    },
    {
      id: "17656bad-07b6-4fd4-bb9a-ccbc54e5ea0a",
      firstName: "Sarah",
      middleName: "Sally",
      lastName: "Samuels",
      email: "sarah@example.com",
      status: "ACTIVE",
      __typename: "ApiUserWithStatus",
    },
    {
      id: "17656bad-08b6-4fd4-bb9a-ccac54e5ea0a",
      firstName: "Nicole",
      middleName: "Suspended",
      lastName: "Carter",
      email: "nicole@example.com",
      status: "SUSPENDED",
      __typename: "ApiUserWithStatus",
    },
  ];
  const store = mockStore(storeData);
  const mocks: MockedResponse[] = [
    {
      request: {
        operationName: "GetUsersAndStatus",
        query: GetUsersAndStatusDocument,
        variables: {},
      },
      result: {
        data: {
          usersWithStatus: mockedUsersWithStatus,
        },
      },
    },
    {
      request: {
        operationName: "GetUser",
        query: GetUserDocument,
        variables: { id: "3a4a221d-a8ca-42b3-aa03-37b93266025b" },
      },
      result: {
        data: {
          user: {
            id: "3a4a221d-a8ca-42b3-aa03-37b93266025b",
            firstName: "Ben",
            middleName: "Billy",
            lastName: "Barnes",
            roleDescription: "Misconfigured user",
            role: null,
            permissions: [],
            email: "ben@example.com",
            status: "ACTIVE",
            organization: {
              testingFacility: [
                {
                  id: "3ea40184-b7d6-4807-b924-747a316a86a6",
                  name: "Testing Site",
                  __typename: "Facility",
                },
              ],
              __typename: "Organization",
            },
            __typename: "User",
          },
        },
      },
    },
  ];

  const supendedUserMocks: MockedResponse[] = [
    {
      request: {
        operationName: "GetUsersAndStatus",
        query: GetUsersAndStatusDocument,
        variables: {},
      },
      result: {
        data: {
          usersWithStatus: mockedUsersWithStatus,
        },
      },
    },
    {
      request: {
        operationName: "GetUser",
        query: GetUserDocument,
        variables: { id: "3a4a221d-a8ca-42b3-aa03-37b93266025b" },
      },
      result: {
        data: {
          user: {
            id: "3a4a221d-a8ca-42b3-aa03-37b93266025b",
            firstName: "Ben",
            middleName: "Billy",
            lastName: "Barnes",
            roleDescription: "Misconfigured user",
            role: null,
            permissions: [],
            email: "ben@example.com",
            status: "ACTIVE",
            organization: {
              testingFacility: [
                {
                  id: "3ea40184-b7d6-4807-b924-747a316a86a6",
                  name: "Testing Site",
                  __typename: "Facility",
                },
              ],
              __typename: "Organization",
            },
            __typename: "User",
          },
        },
      },
    },
    {
      request: {
        operationName: "GetUser",
        query: GetUserDocument,
        variables: { id: "17656bad-08b6-4fd4-bb9a-ccac54e5ea0a" },
      },
      result: {
        data: {
          user: {
            id: "17656bad-08b6-4fd4-bb9a-ccac54e5ea0a",
            firstName: "Nicole",
            middleName: "Suspended",
            lastName: "Carter",
            roleDescription: "Standard user",
            role: "Standard user",
            permissions: [],
            email: "nicole@example.com",
            status: "SUSPENDED",
            organization: {
              testingFacility: [
                {
                  id: "3ea40184-b7d6-4807-b924-747a316a86a6",
                  name: "Testing Site",
                  __typename: "Facility",
                },
              ],
              __typename: "Organization",
            },
            __typename: "User",
          },
        },
      },
    },
  ];
  const renderComponentWithMocks = (
    graphqlResponses: MockedResponse[],
    store: Store<unknown, AnyAction>
  ) =>
    render(
      <MemoryRouter>
        <MockedProvider mocks={graphqlResponses}>
          <Provider store={store}>
            <ManageUsersContainer />
          </Provider>
        </MockedProvider>
      </MemoryRouter>
    );

  it("loads the component and displays users successfully", async () => {
    const { container } = renderComponentWithMocks(mocks, store);
    await waitForElementToBeRemoved(screen.queryByText("Loading..."));
    await waitFor(() =>
      expect(
        screen.queryByRole("heading", {
          level: 2,
          description: /barnes, ben billy/i,
        })
      )
    );
    await screen.findByText(/ben@example.com/i);
    expect(container).toMatchSnapshot();
    expect(document.title).toEqual("Manage users | SimpleReport");
  });

  it("loads the component and user not found ", async () => {
    const notFoundMock = [{ ...mocks[0], result: {} }];
    renderComponentWithMocks(notFoundMock, store);
    await screen.findByText(/Error: Users not found/i);
  });

  it("when user is an org admin , modal copy reflects the reactivate only flow", async () => {
    renderComponentWithMocks(supendedUserMocks, store);

    const suspendedUser = await screen.findByText("Carter, Nicole Suspended");
    await act(async () => await userEvent.click(suspendedUser));

    const reactivateButton = await screen.findByText("Activate user");
    await act(async () => await userEvent.click(reactivateButton));
    expect(await screen.findByText(ORG_ADMIN_REACTIVATE_COPY));
  });

  it("when user is a site admin , modal copy reflects the reset password flow", async () => {
    const storeWithSiteAdminUser = cloneDeep(storeData);
    storeWithSiteAdminUser.user.isAdmin = true;
    const testMockStore = mockStore(storeWithSiteAdminUser);

    renderComponentWithMocks(supendedUserMocks, testMockStore);

    const suspendedUser = await screen.findByText("Carter, Nicole Suspended");
    await act(async () => await userEvent.click(suspendedUser));

    const reactivateButton = await screen.findByText("Activate user");
    await act(async () => await userEvent.click(reactivateButton));
    expect(await screen.findByText(SITE_ADMIN_REACTIVATE_COPY));
  });
});
