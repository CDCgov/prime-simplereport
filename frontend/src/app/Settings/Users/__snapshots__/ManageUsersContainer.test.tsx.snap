// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageUsersContainer loads the component and displays users successfully 1`] = `
<div>
  <div
    class="prime-container card-container manage-users-card"
  >
    <div
      class="usa-card__header"
    >
      <h1>
        Manage users
      </h1>
      <button
        class="usa-button"
        type="button"
      >
        + Add user
      </button>
    </div>
    <div
      class="usa-card__body"
    >
      <div
        class="grid-row"
      >
        <div
          class="display-block users-sidenav"
        >
          <h2
            class="users-sidenav-header"
          >
            Users
          </h2>
          <div
            class="prime-search-container"
          >
            <form
              class="usa-search usa-search--small prime-search-input display-inline-block"
              role="search"
            >
              <label
                class="usa-sr-only"
                for="search-field-small"
              >
                Search by name
              </label>
              <div>
                <input
                  autocomplete="off"
                  class="usa-input"
                  id="search-field-small"
                  name="search"
                  placeholder="Search by name"
                  style="border-right: 1px solid;"
                  type="search"
                  value=""
                />
                <button
                  class="usa-button"
                  disabled=""
                  style="display: none;"
                  type="submit"
                >
                  <img
                    alt="Search"
                    class="usa-search__submit-icon"
                    src="search--white.svg"
                  />
                </button>
              </div>
            </form>
          </div>
          <nav
            aria-label="Tertiary navigation"
            class="prime-secondary-nav maxh-tablet-lg overflow-y-scroll"
          >
            <div
              aria-owns="user-tab-3a4a221d-a8ca-42b3-aa03-37b93266025b user-tab-1029653e-24d9-428e-83b0-468319948902 user-tab-17656bad-08b6-4fd4-bb9a-ccac54e5ea0a user-tab-60bb9e3a-fe8a-4b81-b894-b01649c95e70 user-tab-0d3fa224-3d56-4382-b89c-9d8e415e59b3 user-tab-17656bad-07b6-4fd4-bb9a-ccbc54e5ea0a"
              class="usa-sidenav"
              role="tablist"
            >
              <div
                class="usa-sidenav__item users-sidenav-item"
              >
                <button
                  aria-label="Barnes, Ben Billy"
                  aria-selected="true"
                  class="usa-button--unstyled text-ink text-no-underline padding-105 padding-right-2 padding-left-3 usa-current"
                  id="user-tab-3a4a221d-a8ca-42b3-aa03-37b93266025b"
                  role="tab"
                  style="cursor: pointer;"
                >
                  <div
                    class="sidenav-user-name"
                  >
                    Barnes, Ben Billy
                  </div>
                  <span
                    class="sidenav-user-status padding-left-0"
                  />
                </button>
              </div>
              <div
                class="usa-sidenav__item users-sidenav-item"
              >
                <button
                  aria-label="Bobberoo, Bob"
                  aria-selected="false"
                  class="usa-button--unstyled text-ink text-no-underline padding-105 padding-right-2 padding-left-3"
                  id="user-tab-1029653e-24d9-428e-83b0-468319948902"
                  role="tab"
                  style="cursor: pointer;"
                >
                  <div
                    class="sidenav-user-name"
                  >
                    Bobberoo, Bob
                  </div>
                  <span
                    class="sidenav-user-status padding-left-0"
                  />
                </button>
              </div>
              <div
                class="usa-sidenav__item users-sidenav-item"
              >
                <button
                  aria-label="Carter, Nicole Suspended"
                  aria-selected="false"
                  class="usa-button--unstyled text-ink text-no-underline padding-105 padding-right-2 padding-left-3"
                  id="user-tab-17656bad-08b6-4fd4-bb9a-ccac54e5ea0a"
                  role="tab"
                  style="cursor: pointer;"
                >
                  <div
                    class="sidenav-user-name"
                  >
                    Carter, Nicole Suspended
                  </div>
                  <svg>
                    account-deactivated.svg
                  </svg>
                  <span
                    class="sidenav-user-status"
                  >
                    Account deactivated
                  </span>
                </button>
              </div>
              <div
                class="usa-sidenav__item users-sidenav-item"
              >
                <button
                  aria-label="Jackson, Jamar Donald"
                  aria-selected="false"
                  class="usa-button--unstyled text-ink text-no-underline padding-105 padding-right-2 padding-left-3"
                  id="user-tab-60bb9e3a-fe8a-4b81-b894-b01649c95e70"
                  role="tab"
                  style="cursor: pointer;"
                >
                  <div
                    class="sidenav-user-name"
                  >
                    Jackson, Jamar Donald
                  </div>
                  <span
                    class="sidenav-user-status padding-left-0"
                  />
                </button>
              </div>
              <div
                class="usa-sidenav__item users-sidenav-item"
              >
                <button
                  aria-label="Reynolds, Ruby Raven"
                  aria-selected="false"
                  class="usa-button--unstyled text-ink text-no-underline padding-105 padding-right-2 padding-left-3"
                  id="user-tab-0d3fa224-3d56-4382-b89c-9d8e415e59b3"
                  role="tab"
                  style="cursor: pointer;"
                >
                  <div
                    class="sidenav-user-name"
                  >
                    Reynolds, Ruby Raven
                  </div>
                  <span
                    class="sidenav-user-status padding-left-0"
                  />
                </button>
              </div>
              <div
                class="usa-sidenav__item users-sidenav-item"
              >
                <button
                  aria-label="Samuels, Sarah Sally"
                  aria-selected="false"
                  class="usa-button--unstyled text-ink text-no-underline padding-105 padding-right-2 padding-left-3"
                  id="user-tab-17656bad-07b6-4fd4-bb9a-ccbc54e5ea0a"
                  role="tab"
                  style="cursor: pointer;"
                >
                  <div
                    class="sidenav-user-name"
                  >
                    Samuels, Sarah Sally
                  </div>
                  <span
                    class="sidenav-user-status padding-left-0"
                  />
                </button>
              </div>
            </div>
          </nav>
        </div>
        <div
          aria-labelledby="user-tab-3a4a221d-a8ca-42b3-aa03-37b93266025b"
          class="tablet:grid-col padding-left-3 user-detail-column"
          role="tabpanel"
        >
          <div>
            <h2
              class="display-inline-block margin-top-1 margin-bottom-0 user-name-header"
            >
              Barnes, Ben Billy
            </h2>
            <div
              class="user-status-subheader"
            >
              <span
                class="top-user-status padding-left-0"
              />
            </div>
          </div>
          <nav
            aria-label="User action navigation"
            class="prime-secondary-nav margin-top-4 padding-bottom-0"
          >
            <div
              aria-owns="userinformation-tab facility-access-tab-id"
              class="usa-nav__secondary-links prime-nav usa-list"
              role="tablist"
            >
              <div
                class="usa-nav__secondary-item usa-current"
              >
                <button
                  aria-selected="true"
                  class="usa-button--unstyled text-ink text-no-underline"
                  id="userinformation-tab"
                  role="tab"
                >
                  User information
                </button>
              </div>
              <div
                class="usa-nav__secondary-item "
              >
                <button
                  aria-selected="false"
                  class="usa-button--unstyled text-ink text-no-underline"
                  id="facility-access-tab-id"
                  role="tab"
                >
                  Facility access
                </button>
              </div>
            </div>
          </nav>
          <div
            aria-labelledby="userinformation-tab"
            class="padding-left-1"
            role="tabpanel"
          >
            <h3
              class="basic-info-header"
            >
              Basic information
            </h3>
            <div
              class="user-header grid-row flex-row flex-align-center"
            >
              <div>
                <div
                  class="userinfo-subheader"
                >
                  Name
                </div>
                <p
                  class="userinfo-text"
                >
                  Ben Billy Barnes
                </p>
              </div>
              <button
                aria-disabled="false"
                class="usa-button usa-button--outline margin-left-auto margin-bottom-1"
                type="button"
              >
                Edit name
              </button>
            </div>
            <div
              class="user-header grid-row flex-row flex-align-center"
            >
              <div>
                <div
                  class="userinfo-subheader"
                >
                  Email
                </div>
                <p
                  class="userinfo-text"
                >
                  ben@example.com
                </p>
              </div>
              <button
                aria-disabled="false"
                class="usa-button usa-button--outline margin-left-auto margin-bottom-1"
                type="button"
              >
                Edit email
              </button>
            </div>
            <div
              class="sr-separator-line margin-left-neg-4 margin-right-neg-2"
            />
            <h3
              class="user-controls-header"
            >
              User controls
            </h3>
            <div
              class="user-header grid-row flex-row flex-align-center"
            >
              <div
                class="grid-col margin-right-8"
              >
                <div
                  class="userinfo-subheader"
                >
                  Password
                </div>
                <p
                  class="usercontrols-text"
                >
                  Send a link to reset user password. Users must answer a password recovery question to access their account.
                </p>
              </div>
              <button
                aria-disabled="false"
                class="usa-button usa-button--outline margin-left-auto margin-bottom-1"
                type="button"
              >
                Send password reset email
              </button>
            </div>
            <div
              class="user-header grid-row flex-row flex-align-center"
            >
              <div
                class="grid-col"
              >
                <div
                  class="userinfo-subheader"
                >
                  Reset multi-factor authentication (MFA)
                </div>
                <p
                  class="usercontrols-text"
                >
                  Reset user MFA account access settings
                </p>
              </div>
              <button
                aria-disabled="false"
                class="usa-button usa-button--outline margin-left-auto margin-bottom-1"
                type="button"
              >
                Reset MFA
              </button>
            </div>
            <div
              class="user-header grid-row flex-row flex-align-center"
            >
              <div>
                <div
                  class="userinfo-subheader"
                >
                  Delete user
                </div>
                <p
                  class="usercontrols-text"
                >
                  Permanently delete user account and data
                </p>
              </div>
              <button
                aria-disabled="false"
                class="usa-button usa-button--outline margin-left-auto margin-bottom-1"
                type="button"
              >
                Delete user
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="grid-row"
      >
        <nav
          aria-label="Pagination"
          class="usa-pagination"
          role="navigation"
        >
          <ol>
            <li>
              <a
                aria-label="Page 1"
                class="is-active"
                href="/settings/users/1"
              >
                <span>
                  1
                </span>
              </a>
            </li>
          </ol>
        </nav>
      </div>
    </div>
  </div>
</div>
`;
